File: passform2_ws/src/passform_ros-main/passform_util/passform_util/types/bay.py
from typing import Optional

from basyx.aas import model
from basyx.aas.model import base
from passform_util.basyx import sanitize_basyx_string


class Bay(model.Submodel):
    def __init__(self,
        unique_id: str,
        name: str,
        origin: list=[0.0,0.0,0.0],
        parent: model.Identifiable = None,
        description: Optional[base.LangStringSet] = None,
        virtual: bool = False,
        # properties: List[model.Property] = field(default_factory=list, compare=False),
        **kwargs
    ):
        super().__init__(
            identification=model.Identifier(unique_id, model.IdentifierType.CUSTOM),
            id_short=sanitize_basyx_string(name), #replace ' ', '-'
            category = 'passform_bay' if not virtual else 'virtual_bay',
            parent=parent,
            **kwargs)

        if not virtual:
            self.init_geometry(origin)
            self.init_supply()

    def init_geometry(self, origin:list):
        """
        :origin: list with x,y,z
        """
        assert len(origin) == 3, 'origin must be specified with three elements'

        collection = model.SubmodelElementCollectionUnordered(
            id_short='Origin',
            description={
                'en-us':'Description of the bays origin.',
                'de':'Beschreibung des Ursprungs der Bucht.'
            }
        )
        collection.value.add(
            model.Property(
                id_short='x',
                value_type=model.datatypes.Double,
                value=origin[0],
                description={
                    'en-us':'x-position of the origin in m.',
                    'de':'x-Position des Ursprungs in m.'
                }
            )
        )
        collection.value.add(
            model.Property(
                id_short='y',
                value_type=model.datatypes.Double,
                value=origin[1],
                description={
                    'en-us':'y-position of the origin in m.',
                    'de':'y-Position des Ursprungs in m.'
                }
            )
        )
        collection.value.add(
            model.Property(
                id_short='z',
                value_type=model.datatypes.Double,
                value=origin[2],
                description={
                    'en-us':'z-position of the origin in m.',
                    'de':'z-Position des Ursprungs in m.'
                }
            )
        )
        self.submodel_element.add(collection)

    def init_supply(self):
        collection = model.SubmodelElementCollectionUnordered(
            id_short='SupplyStatus',
            description={
                'en-us':'List of all utilities currently active.',
                'de':'Liste aller aktuell aktiven Versorgungen.'
            }
        )
        collection.value.add(
            model.Property(
                id_short='dc_24v',
                value_type=model.datatypes.Boolean,
                value=True,
                description={
                    'en-us':'24V DC.',
                    'de':'24V DC.'
                }
            )
        )
        collection.value.add(
            model.Property(
                id_short='ac_230v',
                value_type=model.datatypes.Boolean,
                value=False,
                description={
                    'en-us':'230V AC.',
                    'de':'230V AC.'
                }
            )
        )
        self.submodel_element.add(collection)

File: passform2_agent_backend/src/core/types/bay.rs
use serde::{Serialize, Deserialize};
use serde_json::{json, Value};
use crate::util::sanitize_id;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct BaySubmodel {
    pub unique_id: String,
    pub name: String,
    pub origin: [f64; 3],
    pub virtual_bay: bool,
}

impl BaySubmodel {
    pub fn new(unique_id: &str, name: &str, origin: [f64; 3], virtual_bay: bool) -> Self {
        Self {
            unique_id: unique_id.to_string(),
            name: sanitize_id(name),
            origin,
            virtual_bay,
        }
    }

    /// Generiert das BaSyx-konforme JSON-Submodell
    pub fn to_basyx_json(&self) -> Value {
        let category = if self.virtual_bay { "virtual_bay" } else { "passform_bay" };
        
        let mut submodel = json!({
            "idShort": self.name,
            "identification": {
                "id": self.unique_id,
                "idType": "CUSTOM"
            },
            "category": category,
            "modelType": "Submodel",
            "submodelElements": []
        });

        // Wenn nicht virtuell, füge Geometrie und Versorgung hinzu
        if !self.virtual_bay {
            let elements = submodel["submodelElements"].as_array_mut().unwrap();
            elements.push(self.create_geometry_collection());
            elements.push(self.create_supply_collection());
        }

        submodel
    }

    fn create_geometry_collection(&self) -> Value {
        json!({
            "idShort": "Origin",
            "modelType": "SubmodelElementCollection",
            "description": [
                {"language": "en-us", "text": "Description of the bays origin."},
                {"language": "de", "text": "Beschreibung des Ursprungs der Bucht."}
            ],
            "value": [
                self.create_property("x", self.origin[0], "x-position of the origin in m.", "x-Position des Ursprungs in m."),
                self.create_property("y", self.origin[1], "y-position of the origin in m.", "y-Position des Ursprungs in m."),
                self.create_property("z", self.origin[2], "z-position of the origin in m.", "z-Position des Ursprungs in m.")
            ]
        })
    }

    fn create_supply_collection(&self) -> Value {
        json!({
            "idShort": "SupplyStatus",
            "modelType": "SubmodelElementCollection",
            "description": [
                {"language": "en-us", "text": "List of all utilities currently active."},
                {"language": "de", "text": "Liste aller aktuell aktiven Versorgungen."}
            ],
            "value": [
                self.create_bool_property("dc_24v", true, "24V DC.", "24V DC."),
                self.create_bool_property("ac_230v", false, "230V AC.", "230V AC.")
            ]
        })
    }

    fn create_property(&self, id: &str, value: f64, desc_en: &str, desc_de: &str) -> Value {
        json!({
            "idShort": id,
            "modelType": "Property",
            "valueType": "double",
            "value": value,
            "description": [
                {"language": "en-us", "text": desc_en},
                {"language": "de", "text": desc_de}
            ]
        })
    }

    fn create_bool_property(&self, id: &str, value: bool, desc_en: &str, desc_de: &str) -> Value {
        json!({
            "idShort": id,
            "modelType": "Property",
            "valueType": "boolean",
            "value": value,
            "description": [
                {"language": "en-us", "text": desc_en},
                {"language": "de", "text": desc_de}
            ]
        })
    }
}